src/ModelContextProtocol/Authentication/AuthorizationDelegatingHandler.cs

```diff
@@ -0,0 +1,184 @@
+
+        _credentialProvider = credentialProvider;
+        
+        // Select first supported scheme as the default
+        _currentScheme = _credentialProvider.SupportedSchemes.FirstOrDefault() ??
+            throw new ArgumentException("Authorization provider must support at least one authentication scheme.", nameof(credentialProvider));
+    }
+
+    /// <summary>
+    /// Sends an HTTP request with authentication handling.
+    /// </summary>
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        if (request.Headers.Authorization == null)
+        {
+            await AddAuthorizationHeaderAsync(request, _currentScheme, cancellationToken).ConfigureAwait(false);
+        }
+
+        var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+        if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
+        {
+            return await HandleUnauthorizedResponseAsync(request, response, cancellationToken).ConfigureAwait(false);
+        }
+
+        return response;
+    }
+
+    /// <summary>
+    /// Handles a 401 Unauthorized response by attempting to authenticate and retry the request.
+    /// </summary>
+    private async Task<HttpResponseMessage> HandleUnauthorizedResponseAsync(
+        HttpRequestMessage originalRequest, 
+        HttpResponseMessage response, 
+        CancellationToken cancellationToken)
+    {
+        // Gather the schemes the server wants us to use from WWW-Authenticate headers
+        var serverSchemes = ExtractServerSupportedSchemes(response);
+
+        // Find the intersection between what the server supports and what our provider supports
+        string? bestSchemeMatch = null;
+
+        // First try to find a direct match with the current scheme if it's still valid
+        string schemeUsed = originalRequest.Headers.Authorization?.Scheme ?? _currentScheme ?? string.Empty;
+        if (!string.IsNullOrEmpty(schemeUsed) && 
+            serverSchemes.Contains(schemeUsed) && 
+            _credentialProvider.SupportedSchemes.Contains(schemeUsed))
+        {
+            bestSchemeMatch = schemeUsed;
+        }
+        else
+        {
+            // Find the first server scheme that's in our supported set
+            bestSchemeMatch = serverSchemes.Intersect(_credentialProvider.SupportedSchemes, StringComparer.OrdinalIgnoreCase).FirstOrDefault();
+
+            // If no match was found, either throw an exception or use default
+            if (bestSchemeMatch is null)
+            {
+                if (serverSchemes.Count > 0)
+                {
+                    throw new InvalidOperationException(
+                        $"No matching authentication scheme found. Server supports: [{string.Join(", ", serverSchemes)}], " +
+                        $"Provider supports: [{string.Join(", ", _credentialProvider.SupportedSchemes)}].");
```

Comment:
```suggestion
                    throw new IOException(
                        $"The server does not support any of the provided authentication schemes."
                        $"Server supports: [{string.Join(", ", serverSchemes)}], " +
                        $"Provider supports: [{string.Join(", ", _credentialProvider.SupportedSchemes)}].");
```


I think most of the new InvalidOperationExceptions should be IOExceptions instead. As a general rule of thumb, I ask myself is it impossible for me as a developer to write the appropriate if checks to avoid the exception? If there's no way to reliably avoid the exception, it probably shouldn't be an InvalidOperationException.

@stephentoub I know you initially suggested InvalidOperationExceptions, so feel free to disagree, but I assume the main point of that comment was just to avoid defining new Exception types.

Reply:
> I assume the main point of that comment was just to avoid defining new Exception types.

Yup